File: java/solutions/src/main/java/io/github/ajoz/workshop/fp/part2/exercise3/Exercise3.java
package io.github.ajoz.workshop.fp.part2.exercise3;

import io.github.ajoz.workshop.fp.tools.Function2;
import io.github.ajoz.workshop.fp.tools.collections.Lists;

import java.util.List;

class Exercise3 {
    // Part 1:
    static <A, B> B foldLeft(final List<A> list,
                             final B initial,
                             final Function2<B, A, B> operator) {
        B accumulator = initial;
        for (final A element : list) {
            accumulator = operator.apply(accumulator, element);
        }

        return accumulator;
    }

    // Part 2:
    static <A, B> B foldRight(final List<A> list,
                              final B initial,
                              final Function2<A, B, B> operator) {
        B accumulator = initial;
        for (int i = list.size(); i > 0; i--) {
            accumulator = operator.apply(list.get(i - 1), accumulator);
        }
        return accumulator;
    }

    // Part 3:
    static <A, B> B foldRight2(final List<A> list,
                               final B initial,
                               final Function2<A, B, B> operator) {
        return foldLeft(
                Lists.reverse(list),
                initial,
                operator.flip()
        );
    }
}

File: kotlin/tools/src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/part_2/solutions/exercise_3/Exercise3.kt
@file:Suppress("PackageName")

package io.github.ajoz.workshop.fp.kotlin.part_2.solutions.exercise_3

import io.github.ajoz.workshop.fp.java.tools.collections.Lists
import io.github.ajoz.workshop.fp.kotlin.tools.flipped

// Part 1:
fun <A, B> foldLeft(list: List<A>,
                    initial: B,
                    operator: (B, A) -> B): B {
    var accumulator = initial
    for (element in list) {
        accumulator = operator(accumulator, element)
    }

    return accumulator
}

// Part 2:
fun <A, B> foldRight(list: List<A>,
                     initial: B,
                     operator: (A, B) -> B): B {
    var accumulator = initial
    for (i in list.size downTo 1) {
        accumulator = operator(list[i - 1], accumulator)
    }
    return accumulator
}

// Part 3:
fun <A, B> foldRight2(list: List<A>,
                      initial: B,
                      operator: (A, B) -> B): B {
    return foldLeft(
            Lists.reverse(list),
            initial,
            operator.flipped()
    )
}